design/marketplace.md

```diff
@@ -0,0 +1,279 @@
+                                  burned                               burned
+
+
+
+            ---------------- time ---------------->
+
+
+The time interval that a slot is filled by a host determines the host payout;
+for every second of the interval a certain amount of tokens are awarded to the
+host. Hosts that fill a slot are required to submit frequent proofs of storage.
+
+When a proof is missed, the collateral associated with a slot is used to pay a
+fee to the one who marked the proof as missing. The rest of the slot collateral
+is reserved for repairs. The slot is now considered empty again until another
+host submits a correct proof together with collateral. Payouts for the time
+interval that a slot is empty are burned.
+
+Contract lifecycle
+------------------
+
+A contract starts when all slots are filled. Regular storage proofs will be
+required from the hosts that filled the slots.
+
+Some contracts may not attract the required amount of hosts, for instance
+because the payment is insufficient or the storage demands on the network are
+too high. To ensure that such contracts end, we add a timeout to the contract.
+If the contract failed to attract sufficient hosts before the timeout is
+reached, it is considered cancelled, and the hosts that filled any of the slots
+are able to withdraw their collateral. They are also paid for the time interval
+before the timeout. The client is able to withdraw the rest of the tokens in the
+contract.
+
+A contract ends when the money that was paid upfront runs out. The end time can
+be calculated from the amount of tokens that are paid out per second. Note that
+in our scheme this amount does not change during the lifetime of the contract,
+ even when proofs are missed and repair happens. This is a desirable property
+for hosts; they can be sure of a steady source of income, and a predetermined
+contract length. When a contract ends, the hosts may withdraw their collateral.
+
+When too many hosts fail to submit storage proofs, and no other hosts take over
+the slots that they vacate, then the content can be considered lost. The
+contract is considered failed. The collateral of every host in the contract is
+burned as an additional incentive for the network hosts to avoid this scenario.
+The client is able to retrieve any funds that are left in the contract.
+
+        |
+        | create
+        |
+        v
+    -----------       timeout       -------------
+    | created | ------------------> | cancelled |
+    -----------                     -------------
+        |
+        | all slots filled
+        |
+        v
+    -----------   too many losses    ----------
+    | started | -------------------> | failed |
+    -----------                      ----------
+        |
+        | money runs out
+        |
+        v
+    ---------
+    | ended |
+    ---------
+
+
+Repairs
+-------
+
+When a slot becomes empty, the remaining collateral associated with the slot is
```

Comment:
I'm all for getting this amount to be as close as possible to the actual costs of repair. I just don't know any simple way of approximating these costs.

The download receipts sound complex to me for several reasons:
1. They introduce a tight coupling between the storage contracts and the bandwidth incentives. 
2. Payments are handled off-chain in channels, contracts are handled on-chain, which means we have to figure out how to present these payment on-chain.
3.  It seems easy to artificially inflate the price of downloads using a sybil attack whereby you sell the data to yourself. Preventing this seems non-trivial.

Perhaps a fixed price for repair that is set at contract start is at the sweet spot of being both simple and effective enough?

Reply:
Actually, it makes even more sense in that new design. Otherwise, hosts can set their bandwidth price very high, and since the client doesn't choose the host, they could hold the data hostage

Reply:
> I'm all for getting this amount to be as close as possible to the actual costs of repair. I just don't know any simple way of approximating these costs.

This should be doable. We can quantify both the bandwidth and repair costs and we've been looking at this to some extent already - @cskiraly.

> The download receipts sound complex to me for several reasons:
> They introduce a tight coupling between the storage contracts and the bandwidth incentives.
> Payments are handled off-chain in channels, contracts are handled on-chain, which means we have to figure out how to present these payment on-chain.

That should be fine? Ultimately, payment channels are settled on chain and the receipts would be verifiable on-chain as well. It should be just a call to the payment channel contract?

> It seems easy to artificially inflate the price of downloads using a sybil attack whereby you sell the data to yourself. Preventing this seems non-trivial.

Not sure what a sybil attack has to do with this? Every entity is bounded by a stake. An entity attempting to subvert the system by creating many identities exposes itself to (at least) a proportional amount of risk to the number of entities it brings to the system.

Never the less, it's a good point and perhaps you are right that we should set a fixed price for repair.

> Perhaps a fixed price for repair that is set at contract start is at the sweet spot of being both simple and effective enough?

Reply:
> If I may ramble about my old ideas, setting a maximum retrieval price in the contract would, amongst other benefits, allow us to calculate easily how much the repair costed in bandwidth

This doesn't sounds like a bad idea and I think it makes sense, but then I also like the idea of simply fixing the repair price to some average bandwidth cost + resource cost. It simplifies things a lot. 

Furthermore, I think in the case of repair, fixing the price is more acceptable because of several reasons - a) the repair node is going to further benefit from earning on the contract and b) it will also potentially benefit from download fees - so, there is a natural incentive to do the repair.

Reply:
There are two scenarios to watch out for here.

1. If download prices can be inflated, then nodes will attempt to collude to collect the outstanding collateral
2. If the price of repair is much larger than the potential benefits from storing the data, then no one is going to attempt it

So, whatever we choose to do, it should be somewhere in between those two extremes - nodes should not be able to collect the entire collateral during repair and the repair process should not be more expensive than the rewards from storing the data.

Again, a fixed price for repair that averages the costs for bandwidth + resources doesn't sound like a bad idea.

The more interesting issue is what happens if nodes choose not to participate in the repair? I tend to look at this as follows:

- If the redundancy levels are close to optimal, then repair can be lazy and at the same time there should be enough nodes, with enough pieces sufficient to do the repair
- If redundancy levels decrease and are getting close to `r0`, then nodes are incentivized to actually help with repair as much as possible, otherwise, once data is lost, everybody looses the stake

`r0` is a per dataset parameter that is somewhere in between `K` - the minimum number of pieces required to reconstruct and `N` - all the available pieces.

Reply:
Ok, all things considered I think that having a fixed price for repairs is the optimal solution for now. I changed the wording in the proposal to reflect this.